File: PO_Projekt2/ProgOb_Projekt2/src/main/java/pl/Dybowski/Adrian/Projekt2/WirtualnySwiat/Zwierzeta/Czlowiek.java
package pl.Dybowski.Adrian.Projekt2.WirtualnySwiat.Zwierzeta;

import pl.Dybowski.Adrian.Projekt2.WirtualnySwiat.Kierunki;
import pl.Dybowski.Adrian.Projekt2.WirtualnySwiat.Organizm;
import pl.Dybowski.Adrian.Projekt2.WirtualnySwiat.RodzajeKomunikatow;
import pl.Dybowski.Adrian.Projekt2.WirtualnySwiat.Stale;
import pl.Dybowski.Adrian.Projekt2.WirtualnySwiat.Swiat;
import pl.Dybowski.Adrian.Projekt2.WirtualnySwiat.Wspolrzedne2D;

public class Czlowiek extends Zwierze {

	private int trwanieZdolnosci, odnowienieZdolnosci;
	private final int maxTrwanieZdolnosci, maxOdnowienieZdolnosci;
	private boolean zdolnoscAktywna;
	
	private Kierunki kierunek;
	
	public Czlowiek(Swiat swiat, Wspolrzedne2D polozenie) {
		super(swiat, polozenie, Stale.sila.get(Czlowiek.class), Stale.inicjatywa.get(Czlowiek.class),
				Stale.szansaRuchu.get(Czlowiek.class), Stale.zasiegRuchu.get(Czlowiek.class),
				Stale.szansaUcieczki.get(Czlowiek.class), Stale.szansaRozmnazania.get(Czlowiek.class));
		
		maxTrwanieZdolnosci = Stale.maxTrwanieZdolnosci;
		maxOdnowienieZdolnosci = Stale.maxOdnowienieZdolnosci;
		
		trwanieZdolnosci = 0;
		odnowienieZdolnosci = 0;
		zdolnoscAktywna = false;
		
		kierunek = Kierunki.stop;
	}
	
	public void setTrwanieZdolnosci(int trwanieZdolnosci) {
	    this.trwanieZdolnosci = trwanieZdolnosci;
	}

	public void setOdnowienieZdolnosci(int odnowienieZdolnosci) {
	    this.odnowienieZdolnosci = odnowienieZdolnosci;
	}

	public void setZdolnoscAktywna(boolean zdolnoscAktywna) {
	    this.zdolnoscAktywna = zdolnoscAktywna;
	}

	public int getTrwanieZdolnosci() {
	    return trwanieZdolnosci;
	}

	public int getOdnowienieZdolnosci() {
	    return odnowienieZdolnosci;
	}

	public boolean isZdolnoscAktywna() {
	    return zdolnoscAktywna;
	}

	
	public void setKierunek(Kierunki kierunek) {
		this.kierunek = kierunek;
	}
	
	@Override
	public boolean odbilAtak(Organizm atakujacy) {
		return zdolnoscAktywna;
	}
	
	@Override
	public void akcja() {
		if (!zyje) return;
		
		if (!swiat.szansaTrafiona(szansaRuchu)) return;
		
		for (int i = 0; i < zasiegRuchu; i++) {
			Wspolrzedne2D cel = swiat.poleWKierunku(polozenie, kierunek);
			
			if (swiat.poleZajete(cel)) {
				kolizja(swiat.organizmNaPolu(cel), true);
				kierunek = Kierunki.stop;
				return;
			}
			
			przemiescSieNaPole(cel);
		}
		
		kierunek = Kierunki.stop;
	}
	
	public void aktywujZdolnosc() {
		if (odnowienieZdolnosci > 0) return;
		if (zdolnoscAktywna) return;
		
		trwanieZdolnosci = maxTrwanieZdolnosci;
		zdolnoscAktywna = true;
		
		swiat.dodajKomunikat(this, RodzajeKomunikatow.aktywujeMoc);
	}
	
	public void odswiezZdolnosc() {
		if (trwanieZdolnosci > 0) trwanieZdolnosci--;

		if (trwanieZdolnosci == 0 && zdolnoscAktywna == true) {
			zdolnoscAktywna = false;
			odnowienieZdolnosci = maxOdnowienieZdolnosci;
			return;
		}

		if (odnowienieZdolnosci > 0) odnowienieZdolnosci--;
	}
	
	@Override
	public String toString() {
		return this.getClass().getSimpleName()+" "+polozenie.getX()+" "+polozenie.getY()+" "+Integer.toString(getWiek())+" "+sila+" "+trwanieZdolnosci+" "+odnowienieZdolnosci+" "+zdolnoscAktywna+"\n";
	}

	@Override
	public boolean tegoSamegoTypu(Organizm inny) {
		return inny instanceof Czlowiek;
	}

	@Override
	public Organizm stworzDziecko(Wspolrzedne2D cel) {
		return null;
	}

}

File: projekty/PO_Projekt3/czlowiek.py
from zwierze import Zwierze
from kierunki import Kierunki
import random

class Czlowiek(Zwierze):
    def __init__(self, swiat, polozenie):
        super().__init__(swiat, polozenie, 5, 4, 1.00, 1, 0.00, 0.00)
        self.__MAX_TRWANIE_ZDOLNOSCI = 5
        self.__MAX_ODNOWIENIE_ZDOLNOSCI = 5
        self.__trwanieZdolnosci = 0
        self.__odnowienieZdolnosci = 0
        self.__zdolnoscAktywna = False
        self.__kierunek = Kierunki.stop

    def getMaxTrwanieZdolnosci(self):
        return self.__MAX_TRWANIE_ZDOLNOSCI

    def getMaxOdnowienieZdolnosci(self):
        return self.__MAX_ODNOWIENIE_ZDOLNOSCI

    def getTrwanieZdolnosci(self):
        return self.__trwanieZdolnosci

    def setTrwanieZdolnosci(self, trwanieZdolnosci):
        self.__trwanieZdolnosci = trwanieZdolnosci

    def getOdnowienieZdolnosci(self):
        return self.__odnowienieZdolnosci

    def setOdnowienieZdolnosci(self, odnowienieZdolnosci):
        self.__odnowienieZdolnosci = odnowienieZdolnosci

    def odbilAtak(self, atakujacy):
        return self.__zdolnoscAktywna

    def getZdolnoscAktywna(self):
        return self.__zdolnoscAktywna

    def setZdolnoscAktywna(self, zdolnoscAktywna):
        self.__zdolnoscAktywna = zdolnoscAktywna

    def getKierunek(self):
        return self.__kierunek

    def setKierunek(self, kierunek):
        self.__kierunek = kierunek

    def akcja(self):
        if not self._zyje:
            return

        if not random.random() < self._szansaRuchu:
            return

        for i in range(self._zasiegRuchu):
            cel = self._swiat.poleWKierunku(self._polozenie, self.__kierunek)

            if self._swiat.poleZajete(cel):
                self.kolizja(self._swiat.organizmNaPolu(cel), True)
                return

            self.przemiescSieNaPole(cel)

        self.__kierunek = Kierunki.stop

    def aktywujZdolnosc(self):
        if self.__odnowienieZdolnosci > 0:
            return

        if self.__zdolnoscAktywna:
            return

        self.__trwanieZdolnosci = self.__MAX_TRWANIE_ZDOLNOSCI
        self.__zdolnoscAktywna = True
        self._swiat.dodajKomunikat(str(self._polozenie) + ": " + self.nazwaKlasy() + " aktywuje zdolnosc")

    def odswiezZdolnosc(self):
        if self.__trwanieZdolnosci > 0:
            self.__trwanieZdolnosci -= 1

        if self.__trwanieZdolnosci == 0 and self.__zdolnoscAktywna == True:
            self.__zdolnoscAktywna = False
            self.__odnowienieZdolnosci = self.__MAX_ODNOWIENIE_ZDOLNOSCI
            return

        if self.__odnowienieZdolnosci > 0:
            self.__odnowienieZdolnosci -= 1

    def tegoSamegoTypu(self, inny):
        return isinstance(inny, Czlowiek)

    def stworzDziecko(self, cel):
        return None     # czlowiek sie nie rozmnaza

    def __str__(self):
        return f"{self.__class__.__name__} {self._polozenie.x} {self._polozenie.y} {self._wiek} {self._sila} {self.__trwanieZdolnosci} {self.__odnowienieZdolnosci} {self.__zdolnoscAktywna}\n"

    def rysowanie(self):
        return (255, 0, 0)
